vote/src/vote_account.rs

```diff
@@ -309,23 +310,53 @@ impl From<VoteAccount> for AccountSharedData {
 }
 
 impl TryFrom<AccountSharedData> for VoteAccount {
-    type Error = Error;
-    fn try_from(account: AccountSharedData) -> Result<Self, Self::Error> {
-        let vote_account = VoteAccountInner::try_from(account)?;
-        Ok(Self(Arc::new(vote_account)))
-    }
-}
-
-impl TryFrom<AccountSharedData> for VoteAccountInner {
     type Error = Error;
     fn try_from(account: AccountSharedData) -> Result<Self, Self::Error> {
         if !solana_sdk::vote::program::check_id(account.owner()) {
             return Err(Error::InvalidOwner(*account.owner()));
         }
-        Ok(Self {
-            vote_state: VoteState::deserialize(account.data()).map_err(Error::InstructionError)?,
-            account,
-        })
+
+        // Allocate as Arc<MaybeUninit<VoteAccountInner>> so we can initialize in place.
+        let mut inner = Arc::new(MaybeUninit::<VoteAccountInner>::uninit());
+        let inner_ptr = Arc::get_mut(&mut inner)
+            .expect("we're the only ref")
+            .as_mut_ptr();
+
+        // Safety:
+        // - All the addr_of_mut!(...).write(...) calls are valid since we just allocated and so
+        // the field pointers are valid.
+        // - We use write() so that the old values aren't dropped since they're still
+        // uninitialized.
+        unsafe {
+            let vote_state = addr_of_mut!((*inner_ptr).vote_state);
+            // Safety:
+            // - vote_state is non-null and MaybeUninit<VoteState> is guaranteed to have same layout
+            // and alignment as VoteState.
+            // - Here it is safe to create a reference to MaybeUninit<VoteState> since the value is
+            // aligned and MaybeUninit<T> is valid for all possible bit values.
+            let vote_state = &mut *(vote_state as *mut MaybeUninit<VoteState>);
+
+            // Try to deserialize in place
+            if let Err(e) = VoteState::deserialize_into_uninit(account.data(), vote_state) {
```

Comment:
Why not pass `*mut VoteState` directly into `VoteState::deserialize_into_ptr`? Why create a `&mut MaybeUninit<VoteState>` and then convert back to `*mut VoteState` inside `deserialize_into_uninit`?

Reply:
Because deserialize_into_ptr is private, and considering that VoteState seems to be public API, I'm not sure it makes sense to make it public for everyone